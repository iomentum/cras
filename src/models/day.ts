/*

         Day

         / \
     worked holy

*/
export class Day {
  date: Date;

  constructor(date:Date){
    this.date = date;
  }

  totalWorked():number {return 0;}
  getDayDate():Date {return this.date}

}

export class WorkedDay extends Day {
  morning: boolean;
  afternoon: boolean;

  constructor(date: Date) {
    super(date);
    this.morning = false;
    this.afternoon = false;
  }

  addMorning(): void {
    this.morning = true;
  }

  removeMorning(): void {
    this.morning = false;
  }

  addAfternoon(): void {
    this.afternoon = true;
  }

  removeAfternoon(): void {
    this.afternoon = false;
  }

  addWholeDay(): void {
    this.addMorning();
    this.addAfternoon();
  }

  reset(): void {
    this.removeMorning();
    this.removeAfternoon();
  }

  totalWorked(): number {
    return (this.morning ? 0.5 : 0) + (this.afternoon ? 0.5 : 0);
  }
}
// un jour peut etre travaillable ou pas.
// si il est travaillable, une personne peut travailler: toute la journee, une demi journee, ou pas.

export class Holiday extends Day {}