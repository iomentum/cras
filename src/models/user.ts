export class User {
  firstName: string;
  lastName:string;
  customer: string;
  email: string;
  uid: string;
  signatureURL: string;

  constructor (firstName:string,lastName:string,customer: string,email: string,uid: string) {
    this.firstName = firstName;
    this.lastName = lastName;
    this.customer = customer;
    this.email = email;
    this.uid = uid;
    this.signatureURL = '';
  }

  setInfos (firstName:string, lastName:string, customer: string, email:string, uid:string): void {
    this.firstName = firstName;
    this.lastName = lastName;
    this.customer = customer;
    this.email = email;
    this.uid = uid;
  }

  setSignatureURL(url:string): void {
    this.signatureURL = url;
  }

  setFirstName(username:string): void {
    this.firstName = username;
  }

  setCustomer(customer:string): void {
    this.customer = customer;
  }

  resetSignatureURL(): void {
    this.signatureURL = "";
  }
}