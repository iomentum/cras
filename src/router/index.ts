import { createWebHistory, createRouter } from "vue-router";
import MainView from '@/views/MainView.vue';
import PrintView from '@/views/PrintView.vue';

const routes = [
  {
    path: "/",
    name: "Home",
    component: MainView
  },
  {
    path: "/print",
    name: "PrintView",
    component: PrintView
  }
];

const router = createRouter({
  history: createWebHistory(),
  routes
});

export default router;
