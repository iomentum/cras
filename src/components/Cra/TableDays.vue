<script setup lang="ts">
import { computed } from '@vue/reactivity';
import { Day, WorkedDay } from '@/models/day';
import SingleDay from '@/components/Cra/SingleDay.vue';
import { useDaysStore } from '@/stores/daysStore';
import { useRoute } from 'vue-router';

const route = useRoute();
const store = useDaysStore();
const daysList = computed(():Day[] | [] => store.getDays);

const toggleAllDays = (evt: Event) => {
  const target = evt.target as HTMLInputElement;
  store.toggleAllDays(target.checked);
};

const isAllDaysChecked = computed(():boolean => {
  const filteredDays: WorkedDay[] = daysList.value.filter((day):boolean => day instanceof WorkedDay) as WorkedDay[];
  const areAllDaysWorked = filteredDays.every(day => day.morning && day.afternoon);

  return areAllDaysWorked;
});

const isMainView = computed(() => {
  return route.path == "/";
});

</script>

<template>
  <div class="table-of-days">
    <div class="left-column">
      <div>Jours</div>
      <div v-if="isMainView">
        <p>
          Jours travaillés <br>
          <input
            v-model="isAllDaysChecked"
            type="checkbox"
            @change="toggleAllDays"
          >
        </p>
      </div>
      <div v-else>
        <p>Jours travaillés</p>
      </div>
    </div>

    <div class="days-container">
      <SingleDay
        v-for="day in daysList"
        :key="day.date.getDay"
        :day="day"
      />
    </div>
    <div class="right-column">
      <div>TOTAL</div>
      <div><p>{{ store.getTotalWorked }}</p></div>
    </div>
  </div>
</template>

<style lang="scss">

.table-of-days {
  font-size: 1rem;
  z-index: 3;
  display: flex;
  text-align: center;
  margin: 0 auto;
  width: 90%;
  font-family: "bau-bold", Arial, Helvetica, sans-serif;
  color: black;

  & .days-container {
    display: flex;
    flex-direction: row;

    & div {
      text-align: center;
    }
  }

  & .left-column {
    border: 1px solid $main-color;
    font-family: "Bau-regular", Arial, Helvetica, sans-serif;
    border-right: 0;

    & p {
      margin: 0 5px;
    }

    & div:first-child {
      font-family: "bau-bold", Arial, Helvetica, sans-serif;
      border-bottom: 1px solid $main-color;
      padding: 5px 4px;
      min-width: 20px;
      font-size: 1rem;
      color: $main-color;
    }

  }

  & .right-column {
    border: 1px solid $main-color;

    & div:nth-of-type(1) {
      padding: 5px 4px;
      font-size: 1rem;
      border-bottom: 1px solid $main-color;
      background-color: $main-color;
      color: white;
    }
  }
}
</style>
