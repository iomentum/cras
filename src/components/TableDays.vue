<script setup lang="ts">
import { computed, ref } from '@vue/reactivity';
import { Day, WorkedDay } from '@/models/day';
import SingleDay from '@/components/SingleDay.vue';
import { useDaysStore } from '@/stores/daysStore';
import { useRoute } from 'vue-router';

const route = useRoute();
const store = useDaysStore();
const daysList = computed(():Day[] | [] => store.getDays);

const toggleAllDays = (evt: Event) => {
  const target = evt.target as HTMLInputElement;
  store.toggleAllDays(target.checked);
};

const isAllDaysChecked = computed(():boolean => {
  const filteredDays: WorkedDay[] = daysList.value.filter((day):Boolean => day instanceof WorkedDay);
  const areAllDaysWorked = filteredDays.every(day => day.morning && day.afternoon);
  return areAllDaysWorked
});

const isMainView = () => route.path == "/"

</script>

<template>
  <div class="table-of-days">
    <div class="left-column">
      <div>Jours</div>
      <div v-if="isMainView()">
        Jours travaillés
        <input
          type="checkbox"
          v-model="isAllDaysChecked"
          @change="toggleAllDays"
        >
      </div>
      <div v-else>
        <p>Jours travaillés</p>
      </div>
    </div>

    <div class="days-container">
      <SingleDay
        v-for="day in daysList"
        :key="day.date.getDay"
        :day="day"
      />
    </div>
    <div class="right-column">
      <div>TOTAL</div>
      <div><p>{{ store.getTotalWorked }}</p></div>
    </div>
  </div>
</template>

<style lang="scss">

.table-of-days {
  z-index: 3;
  display: flex;
  text-align: center;
  height: 90px;
  margin: 0 auto;
  width: 90%;
  font-family: "bau-bold", Arial, Helvetica, sans-serif;
  color: black;

  & .days-container {
    display: flex;
    flex-direction: row;

    & div {
      text-align: center;
    }
  }

  & .left-column {
    border: 1px solid $main-color;
    font-family: "Bau-regular", Arial, Helvetica, sans-serif;
    border-right: 0;

    & div {
      width: 150px;
    }

    & div:first-child {
      font-family: "bau-bold", Arial, Helvetica, sans-serif;
      border-bottom: 1px solid $main-color;
      height: 40%;
      color: $main-color;
    }

  }

  & .right-column {
    border: 1px solid $main-color;

    & div:nth-of-type(1) {
      border-bottom: 1px solid $main-color;
      height: 40%;
      background-color: $main-color;
      color: white;
      width: 80px;
    }

    & div:nth-of-type(2) {
      height: 59%;
    }
  }
}
</style>
