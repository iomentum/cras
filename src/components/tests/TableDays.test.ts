import { mount } from '@vue/test-utils'
import { WorkedDay, Holiday, Day } from '@/models/day'
import TableDays from '@/components/TableDays.vue'
import { setActivePinia, createPinia } from 'pinia';
import { createTestingPinia } from '@pinia/testing'
import { createApp } from 'vue';
import router from '@/router/index';

const arrayOfDays = [
  new Holiday(new Date(2022,4,1)),
  new WorkedDay(new Date(2022,4,2)),
  new WorkedDay(new Date(2022,4,3)),
  new WorkedDay(new Date(2022,4,4)),
  new WorkedDay(new Date(2022,4,5)),
  new WorkedDay(new Date(2022,4,6)),
  new Holiday(new Date(2022,4,7)),
  new Holiday(new Date(2022,4,8)),
  new WorkedDay(new Date(2022,4,9)),
  new WorkedDay(new Date(2022,4,10)),
  new WorkedDay(new Date(2022,4,11)),
  new WorkedDay(new Date(2022,4,12)),
  new WorkedDay(new Date(2022,4,13)),
  new Holiday(new Date(2022,4,14)),
  new Holiday(new Date(2022,4,15)),
  new WorkedDay(new Date(2022,4,16)),
  new WorkedDay(new Date(2022,4,17)),
  new WorkedDay(new Date(2022,4,18)),
  new WorkedDay(new Date(2022,4,19)),
  new WorkedDay(new Date(2022,4,20)),
  new Holiday(new Date(2022,4,21)),
  new Holiday(new Date(2022,4,22)),
  new WorkedDay(new Date(2022,4,23)),
  new WorkedDay(new Date(2022,4,24)),
  new WorkedDay(new Date(2022,4,25)),
  new Holiday(new Date(2022,4,26)),
  new WorkedDay(new Date(2022,4,27)),
  new Holiday(new Date(2022,4,28)),
  new Holiday(new Date(2022,4,29)),
  new WorkedDay(new Date(2022,4,30)),
  new WorkedDay(new Date(2022,4,31)),
];

describe('TableDays unit test', () => {
  const app = createApp({})
  beforeEach(() => {
    const pinia = createPinia()
    app.use(pinia)
    setActivePinia(pinia)
  });

  const wrapper = mount(TableDays,{
    props: {
      daysList: arrayOfDays
    },
    global: {
      plugins: [createTestingPinia(), router],
    }
  });

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
});
