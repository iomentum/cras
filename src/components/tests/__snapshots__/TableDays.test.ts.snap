// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableDays unit test renders correctly 1`] = `
<div
  class="table-of-days"
  dayslist="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
>
  <div
    class="left-column"
  >
    <div>
      Jours
    </div>
    <div>
       Jours travaill√©s 
      <input
        type="checkbox"
      />
    </div>
  </div>
  <div
    class="days-container"
  >
    
    
  </div>
  <div
    class="right-column"
  >
    <div>
      TOTAL
    </div>
    <div>
      <p>
        0
      </p>
    </div>
  </div>
</div>
`;
